File: works/16956.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int R,C;
char arr[501][501];


int main()
{
    cin >> R >> C;
    for(int i = 0; i < R; i++)
    {
        for(int j = 0; j < C; j++)
        {
            cin >> arr[i][j];
            if(arr[i][j] == '.') arr[i][j] = 'D';
        }
    }

    for(int i = 0; i < R; i++)
    {
        for(int j = 0; j < C; j++)
        {
            if(arr[i][j] != 'W') continue;
            bool flag = false;

            if(i-1 >= 0 && arr[i-1][j] == 'S') flag = true;
            if(j+1 < C && arr[i][j+1] == 'S') flag = true;
            if(i+1 < R && arr[i+1][j] == 'S') flag = true;
            if(j-1 >= 0 && arr[i][j-1] == 'S') flag = true;

            if(flag)
            {
                cout << 0;
                return 0;
            }
        }
    }

    cout << 1 << '\n';
    for(int i = 0; i < R; i++)
    {
        for(int j = 0; j < C; j++)
        {
            cout << arr[i][j];
        } cout << '\n';
    }

}

File: works/16956.kt
import java.util.*;


fun main(args: Array<String>)
{
    val sc = Scanner(System.`in`)
    var R = sc.nextInt()
    var C = sc.nextInt()
    var arr = Array(501){CharArray(501){'a'}}
    sc.nextLine()

    for(i in 0 until R)
    {
        var r = sc.nextLine()
        for(j in 0 until C)
        {
            arr[i][j] = r[j]
        }
    }

    for(i in 0 until R)
    {
        for(j in 0 until C)
        {
            if(arr[i][j] != 'W') continue
            var flag = false

            if(i-1 >= 0)
                if(arr[i-1][j] == 'S') flag = true
            if(j+1 < C)
                if(arr[i][j+1] == 'S') flag = true
            if(i+1 < R)
                if(arr[i+1][j] == 'S') flag = true
            if(j-1 >= 0)
                if(arr[i][j-1] == 'S') flag = true

            if(flag)
            {
                print(0)
                return
            }
        }
    }

    println(1)
    for(i in 0 until R)
    {
        for(j in 0 until C)
        {
            if(arr[i][j] == '.') print('D')
            else print(arr[i][j])
        }
        println()
    }

}